Enable AES encryption on Browser WASM

eerhardt opened the pull request:
This ports the C# implementation of AES from https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/mscorlib/system/security/cryptography/rijndaelmanagedtransform.cs, marks the APIs as supported on Browser, and enables AES tests on Browser WASM.

Contributes to https://github.com/dotnet/runtime/issues/40074

The AES SubtleCrypto support will come in a separate PR.

It is easiest to review the commits separately. I copied the reference source, get it formatted and compiling. Then enable new functionality and refactor in separate commits to make it easier to review.

eerhardt committed:
Copy RijndaelManagedTransform from reference source

Copied from https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/mscorlib/system/security/cryptography/rijndaelmanagedtransform.cs

eerhardt committed:
Format document

eerhardt committed:
Clean up RijndaelManagedTransform

1. Add copyright
2. Remove unnecessary attributes
3. Mark class as internal
4. Remove FEATURE_CRYPTO define. Only support CipherMode.CBC and PaddingMode.PKCS7.
5. Remove Contracts and replace with Debug.Assert.

eerhardt committed:
Get RijndaelManagedTransform to compile successfully.

eerhardt committed:
Make RijndaelManagedTransform implement ILiteSymmetricCipher

And hook it up to AesImplementation for Browser.

eerhardt committed:
Enable Aes OneShot Encrypt and Decrypt

eerhardt committed:
Enable all the Aes tests on Browser.

Mark Aes APIs as supported on Browser.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1360 @@
+                throw new ArgumentNullException(nameof(rgbKey));
+
+            Debug.Assert(mode == CipherMode.CBC, "CipherMode is unsupported");
+            Debug.Assert(PaddingValue == PaddingMode.PKCS7, "PaddingMode is unsupported");
```
Was this taken from the .NET Framework?

Rijndael in .NET Framework permitted non-standard block sizes. We should `Debug.Assert(blockSize == 16)` (assuming the `blockSize` parameter is passed in bytes. If it's passed in bits, then `Debug.Assert(blockSize == 128)`.

I just want to make sure we don't accidentally start supporting AES with non-AES block sizes.

vcsjones replied:
Answering my own question because reading helps...

> This ports the C# implementation of AES from https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/mscorlib/system/security/cryptography/rijndaelmanagedtransform.cs,


So I think asserting the block size would be a good idea (and in any other blockSize inputs).

eerhardt committed:
Fix build with latest code analysis rules

eerhardt committed:
Mark all CipherModes and PaddingModes other than CBC and PKCS7 as unsupported on Browser

eerhardt commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesCipherOneShotTests.cs:
```diff
@@ -166,362 +165,124 @@ public static IEnumerable<object[]> TestCases
-
-                // plaintext requires padding
-                yield return new object[]
+                if (PlatformDetection.IsNotBrowser)
```
Done. I also am marking all CipherModes other than CBC as unsupported on browser.

eerhardt committed:
Make fields readonly and rename according to code style.

eerhardt committed:
Replace RijndaelManagedTransformMode with bool encrypting.

eerhardt committed:
Throw PNSE for unsupported CipherMode and PaddingMode.

Remove unnecessary parameters and fields that aren't supported.

eerhardt committed:
Fix DecryptorReuse_LeadsToSameResults test so it works on Browser and passes for all CipherModes on other platforms

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1307 @@
+        {
+            _encrypting = encrypting;
+            _Nr = GetNumberOfRounds(key);
+            _Nb = BlockSizeBytes / 4;
```
BlockSizeBytes is a const.  Presumably _Nb can / should be one, too?

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1307 @@
+            _Nb = BlockSizeBytes / 4;
+            _Nk = key.Length / 4;
+
+            int S1 = _Nb > 6 ? 3 : 2;
+            int S2 = _Nb > 6 ? 4 : 3;
```
And once it's a const, so too can be these.

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1307 @@
+            }
+            if (_lastBlockBuffer != null)
+            {
+                Array.Clear(_lastBlockBuffer, 0, _lastBlockBuffer.Length);
```
Array.Clear now has an overload that takes just the array.

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1307 @@
+            int S2 = _Nb > 6 ? 4 : 3;
+
+            // Precompute the modulus operations: these are performance killers when called frequently
+            int[] encryptindex1 = new int[_Nb];
+            int[] encryptindex2 = new int[_Nb];
+            int[] encryptindex3 = new int[_Nb];
+
+            int[] decryptindex1 = new int[_Nb];
+            int[] decryptindex2 = new int[_Nb];
+            int[] decryptindex3 = new int[_Nb];
```
We can allocate _encryptIndex and _decryptIndex, and then these can just be slices over them.  Then you can remove the Array.Copy calls below, and you avoid these additional array allocations.

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1307 @@
+
+        public int Transform(ReadOnlySpan<byte> input, Span<byte> output)
+        {
+            // TODO: eerhardt change this to make it efficient
```
:smile:

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1307 @@
+            return ((x & unchecked((int)0x7f7f7f7f)) << 1) ^ ((u - (u >> 7 & 0x01FFFFFF)) & 0x1b1b1b1b);
+        }
+
+        private static readonly byte[] s_Sbox = new byte[] {
```
```suggestion
        private static ReadOnlySpan<byte> Sbox => new byte[] {
```

eerhardt committed:
Spanify the input buffer in RijndaelManagedTransform Transform and TransformFinal.

eerhardt committed:
First round of PR feedback

eerhardt committed:
Make _Nb constant BlockSizeInts

eerhardt committed:
Get rid of unnecessary allocations

eerhardt committed:
Use correct exception message resources

eerhardt commented:
I believe this is ready for review. It may be easier to review commit-by-commit and skipping the first 2 commits since they are copying the reference source code and CTRL+K+D respectively.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1282 @@
+                decryptindex3[j] = (j - 3 + BlockSizeInts) % BlockSizeInts;
+            }
+
+            // we only support CBC mode
+            if (iv.Length / 4 != BlockSizeInts)
+                throw new CryptographicException(SR.Cryptography_InvalidIVSize);
```
I think this would read more naturally as `iv.Length != BlockSizeBytes`.

The comment is somewhat misleading - whether ECB or CBC, we still only want to support AES block sizes. I would change the comment to something like "we only support the standard AES block size".

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RijndaelManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1282 @@
+            // We need to always zeroize the following fields because they contain sensitive data
+            if (_IV != null)
+            {
+                Array.Clear(_IV);
```
I think in the spirit of "if we are going to bother zeroing things, might as well do it the """right""" way" using `CryptographicOperations.ZeroMemory`.

eerhardt replied:
`CryptographicOperations.ZeroMemory` only works on `Span<byte>`. These are `int[]`.

So I could do one of the following:

1. Unsafe cast the `int[]` to a `Span<byte>`.
2. Add a new overload of `CryptographicOperations.ZeroMemory` for `Span<int>`.

I wasn't sure if either of these were "worth it". Thoughts?

vcsjones replied:
Ah, no, you're right. I think it's probably fine to leave alone for now.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoOneShot.cs:
```diff
@@ -20,6 +20,10 @@ internal static class UniversalCryptoOneShot
                 throw new CryptographicException(SR.Cryptography_PartialBlock);
 
+            bool cipherHandlesPadding = cipher.HandlesPadding;
+            if (cipherHandlesPadding)
```
I think most of the changes in this method can be undone based on how the implementation works. I think you can just handle `cipherHandlesPadding` up-front:

```C#
if (cipherHandlesPadding)
{
    cipher.ValidatePaddingMode(paddingMode);
    bytesWritten = cipher.TransformFinal(input, output);
    return true;
}

// The rest of the method
```

The rest of this decryption implementation is about handling various padding scenarios, we we don't have them here since the browser implementation does padding for us.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoOneShot.cs:
```diff
@@ -207,21 +243,31 @@ internal static class UniversalCryptoOneShot
+                bytesWritten = cipher.TransformFinal(input, output);
+                return true;
+            }
+            else
```
Since the `if` exits at the end, do we really need the `else` along with all of the indentation for it? Can the `if` just be an early exit?

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoOneShot.cs:
```diff
@@ -20,6 +20,10 @@ internal static class UniversalCryptoOneShot
                 throw new CryptographicException(SR.Cryptography_PartialBlock);
 
+            bool cipherHandlesPadding = cipher.HandlesPadding;
+            if (cipherHandlesPadding)
```
The rest of the implementation of this class is iffy once webcrypto starts getting involved since it relies on being able to perform individual blocks (`Transform` instead of `TransformFinal`), whereas webcrypto gives you a one-shot.

For the sake of simplicity, I think you could implement it like:

```C#
if (cipherHandlesPadding)
{
    cipher.ValidatePaddingMode(paddingMode);

    try
    {
        bytesWritten = cipher.TransformFinal(input, output);
        return true;
    }
    catch (Exception ex) when (ex is ArgumentException or CryptographicException)
    {
        bytesWritten = 0;
        // Don't leave partially copied data in the output
        CryptographicOperations.ZeroMemory(output);
        return false;
    }
}
```

An alternative might be to not use this class at all - and instead change `AesImplementation.Browser` to not use this one-shot mechanism at all.

eerhardt replied:
> An alternative might be to not use this class at all - and instead change AesImplementation.Browser to not use this one-shot mechanism at all.

I had started down this route, but it felt like I was "fighting the design" too much. I would need to make a bunch of .csproj changes on which .cs files were included, and the existing one would need to be named `*.NotBrowser.cs`, etc. Also, I would need to handle all the cases that `OneShotDecrypt` already handled for me. In the end, it just felt "wrong".

I wasn't sure if the `cipher.HandlesPadding` / `cipher.ValidatePaddingMode` was a great design choice either, but I felt like it fit into the existing model better.

@bartonjs @GrabYourPitchforks - any thoughts here?

vcsjones replied:
I would consider splitting `UniversalCryptoOneShot` then by implementation, then. It's fairly complicated as I've added to it over the past few months trying to eke out more performance.

I think this diff highlights what my suggestion is: https://github.com/dotnet/runtime/commit/7741c51c58d0d58507de3c5d0f3de557fa7eea08

vcsjones replied:
Ideally, `RijndaelManagedTransform.Browser.cs` could be taught how to `TransformFinal` directly to a `ReadOnlySpan<byte>` instead of following the required "return an allocation" pattern. If `DecryptData` and `EncryptData` can write to span instead of "write to a byte array or maybe make one for you if you want" would help and could get rid of some "rent, transform, copy" patterns.

eerhardt replied:
The latest changes have resolved this. PTAL

vcsjones commented:
The more I think about it, the more I wonder how this can be implemented for SubtleCrypto. Best as I can tell, implementing `Transform` on `ICryptoTransform` is not possible. Only `TransformFinal` because the SubtleCrypto implementation is a one-shot.

Unlike with hashes, I don't think the data can be buffered until it is finalized because `Transform`'s contract is to write out blocks as they are transformed. Not being able to implement `Transform` would probably not work for a lot of folks, and would mean things like `CryptoStream` would not work, either.

Am I misunderstanding something?

eerhardt committed:
Remove unsafe pointers from RijndaelManagedTransform.

eerhardt commented:
> how this can be implemented for SubtleCrypto

I haven’t tried it yet, but the idea that @bartonjs had was to remember that last block of output data from the previous block, and then use it as the IV for the next block. SubtleCrypto let’s us specify the key, IV, and data.

So the first Transform, pass in the key, original IV, and data. Then take the last block of cipherText on the way out and store it in a field. The next call to Transform (or TransformFinal) use that fields data as IV.

One other thing we will have to do is strip the full block of PCKS7 padding that will come out on a full Transform block. We won’t want to put padding inbetween the the blocks.

Thinking of that, decrypting blocks might be an issue though. I’m not sure how to handle the padding in that case.

if all else fails, Transform returns “bytes written” to the output buffer. We can just buffer and return 0 for each block. Then return the whole thing on TransformFinal.

vcsjones commented:
> remember that last block of output data from the previous block, and then use it as the IV for the next block. SubtleCrypto let’s us specify the key, IV, and data.

Okay. That makes sense.

eerhardt commented:
> remember that last block of output data from the previous block, and then use it as the IV for the next block. SubtleCrypto let’s us specify the key, IV, and data.

I just tested this out quickly, and it appears to work (on the encryption side). Here's roughly what it would look like to append 2 full blocks together:

```ts
function getMessage() {
  const message = "a".repeat(32);
  const enc = new TextEncoder();
  return enc.encode(message);
}

/*
Encrypt the plainText twice in two separate blocks.
*/
async function encryptMessage(key: CryptoKey) {
  const plainText = getMessage();
  iv = window.crypto.getRandomValues(new Uint8Array(16));

  ciphertext = await window.crypto.subtle.encrypt(
    {
      name: "AES-CBC",
      iv
    },
    key,
    plainText
  );

  // since the message is a whole block and we are using PKCS7, we are guaranteed the whole last block is all padding (all values are `0x10`)
  const ciphertextNoPadding = new Uint8Array(ciphertext).slice(0, ciphertext.byteLength - 16);

  // store the last block of the encrypted message as the IV for the next block
  const nextIv = ciphertextNoPadding.subarray(ciphertextNoPadding.length - 16, ciphertextNoPadding.length);

  // encrypt the same message again
  const ciphertext2 = await window.crypto.subtle.encrypt(
    {
      name: "AES-CBC",
      iv: nextIv
    },
    key,
    plainText
  );

  ciphertext = new ArrayBuffer(ciphertextNoPadding.byteLength + ciphertext2.byteLength);
  const ciphertextBytes = new Uint8Array(ciphertext);

  ciphertextBytes.set(ciphertextNoPadding);
  ciphertextBytes.set(new Uint8Array(ciphertext2), ciphertextNoPadding.byteLength);
}
```

eerhardt committed:
PR feedback

Encrypt and Decrypt index arrays are constant values.
Use CopyTo in more places.

eerhardt committed:
Rename RijndaelManagedTransform => AesManagedTransform

eerhardt committed:
Change AesManagedTransform to no longer do padding.

Instead, reuse the UniversalCrypto infrastructure to do padding.

eerhardt committed:
Enable padding tests again, now that Browser supports all padding modes.

eerhardt commented:
After chatting with @bartonjs, he instructed me on how to handle `PaddingMode.None` on top of SubtleCrypto. This means we can reuse all the "UniversalCrypto" infrastructure, and do all the padding in managed code (like we do for all the other backends).

With this knowledge, I can completely simplify the managed implementation by cutting out the "depad buffer" and the padding logic.

I believe this PR is ready for review. Once this goes in, I can send a new PR for the SubtleCrypto side of things.

eerhardt committed:
Remove UnsupportedPlatform for PaddingMode

eerhardt commented on src/libraries/System.Security.Cryptography/src/Resources/Strings.resx:
```diff
@@ -753,4 +753,10 @@
+  <data name="PlatformNotSupported_CipherModeBrowser" xml:space="preserve">
+    <value>Only CipherMode.CBC is supported on this platform.</value>
+  </data>
+  <data name="PlatformNotSupported_PaddingModeBrowser" xml:space="preserve">
```
Remove this - it is no longer used.

eerhardt commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesImplementation.Apple.cs:
```diff
@@ -7,7 +7,7 @@ namespace System.Security.Cryptography
     internal sealed partial class AesImplementation
     {
-        private static UniversalCryptoTransform CreateTransformCore(
+        private static ICryptoTransform CreateTransformCore(
```
These can be reverted now.

eerhardt commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SymmetricPadding.cs:
```diff
@@ -149,6 +149,8 @@ public static bool DepaddingRequired(PaddingMode padding)
 
         public static int GetPaddingLength(ReadOnlySpan<byte> block, PaddingMode paddingMode, int blockSize)
         {
+            Debug.Assert(!OperatingSystem.IsBrowser() || paddingMode == PaddingMode.PKCS7);
```
Revert.

eerhardt committed:
PR feedback

- clean up unnecessary changes

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1031 @@
+        // resets the state of the transform
+        //
+
+        void ILiteSymmetricCipher.Reset(ReadOnlySpan<byte> iv) => throw new NotImplementedException(); // never invoked
```
hmm, there are some refactoring underway that would hit this, but right now you have no way to exercise this in tests... so... I think this is fine.

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/Symmetric/SymmetricOneShotBase.cs:
```diff
@@ -469,6 +470,7 @@ public void DecryptOneShot_Cbc_InvalidPadding_DoesNotContainPlaintext(PaddingMod
         [InlineData(PaddingMode.PKCS7, 2048)]
         [InlineData(PaddingMode.ANSIX923, 2048)]
         [InlineData(PaddingMode.ISO10126, 2048)]
+        [SkipOnPlatform(TestPlatforms.Browser, "CipherMode.ECB is not supported on Browser")]
```
I see a bunch of suppressions about ECB and CFB not being supported, but no test that verifies the experience.  Please add such a test.

vcsjones replied:
I think those are tested in `AesTests.Browser`, no? https://github.com/dotnet/runtime/blob/e087e9520dc4d5c178e5cdafab2ea5aa487b2061/src/libraries/System.Security.Cryptography/tests/AesTests.Browser.cs

bartonjs replied:
I see that it was just at the other end of the PR.

bartonjs commented on src/libraries/System.Security.Cryptography/ref/System.Security.Cryptography.cs:
```diff
@@ -283,10 +280,14 @@ public sealed partial class ChaCha20Poly1305 : System.IDisposable
     public enum CipherMode
     {
         CBC = 1,
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         ECB = 2,
         [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         OFB = 3,
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         CFB = 4,
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         CTS = 5,
```
Personally, I'd rather not paint the enum members here, and leave it as a gap in the annotations.  If someone... really... wanted to, they could write an AES implementation that worked for some of these modes.

The joys of polymorphism.

bartonjs commented on src/libraries/System.Security.Cryptography/ref/System.Security.Cryptography.cs:
```diff
@@ -2302,39 +2301,61 @@ public abstract partial class SymmetricAlgorithm : System.IDisposable
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public byte[] EncryptCfb(byte[] plaintext, byte[] iv, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.None, int feedbackSizeInBits = 8) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public byte[] EncryptCfb(System.ReadOnlySpan<byte> plaintext, System.ReadOnlySpan<byte> iv, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.None, int feedbackSizeInBits = 8) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public int EncryptCfb(System.ReadOnlySpan<byte> plaintext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.None, int feedbackSizeInBits = 8) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public byte[] EncryptEcb(byte[] plaintext, System.Security.Cryptography.PaddingMode paddingMode) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public byte[] EncryptEcb(System.ReadOnlySpan<byte> plaintext, System.Security.Cryptography.PaddingMode paddingMode) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public int EncryptEcb(System.ReadOnlySpan<byte> plaintext, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode) { throw null; }
         public abstract void GenerateIV();
         public abstract void GenerateKey();
         public int GetCiphertextLengthCbc(int plaintextLength, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.PKCS7) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public int GetCiphertextLengthCfb(int plaintextLength, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.None, int feedbackSizeInBits = 8) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public int GetCiphertextLengthEcb(int plaintextLength, System.Security.Cryptography.PaddingMode paddingMode) { throw null; }
         public bool TryDecryptCbc(System.ReadOnlySpan<byte> ciphertext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, out int bytesWritten, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.PKCS7) { throw null; }
         protected virtual bool TryDecryptCbcCore(System.ReadOnlySpan<byte> ciphertext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, out int bytesWritten) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public bool TryDecryptCfb(System.ReadOnlySpan<byte> ciphertext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, out int bytesWritten, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.None, int feedbackSizeInBits = 8) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         protected virtual bool TryDecryptCfbCore(System.ReadOnlySpan<byte> ciphertext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, int feedbackSizeInBits, out int bytesWritten) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public bool TryDecryptEcb(System.ReadOnlySpan<byte> ciphertext, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, out int bytesWritten) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         protected virtual bool TryDecryptEcbCore(System.ReadOnlySpan<byte> ciphertext, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, out int bytesWritten) { throw null; }
         public bool TryEncryptCbc(System.ReadOnlySpan<byte> plaintext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, out int bytesWritten, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.PKCS7) { throw null; }
         protected virtual bool TryEncryptCbcCore(System.ReadOnlySpan<byte> plaintext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, out int bytesWritten) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public bool TryEncryptCfb(System.ReadOnlySpan<byte> plaintext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, out int bytesWritten, System.Security.Cryptography.PaddingMode paddingMode = System.Security.Cryptography.PaddingMode.None, int feedbackSizeInBits = 8) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         protected virtual bool TryEncryptCfbCore(System.ReadOnlySpan<byte> plaintext, System.ReadOnlySpan<byte> iv, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, int feedbackSizeInBits, out int bytesWritten) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public bool TryEncryptEcb(System.ReadOnlySpan<byte> plaintext, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, out int bytesWritten) { throw null; }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         protected virtual bool TryEncryptEcbCore(System.ReadOnlySpan<byte> plaintext, System.Span<byte> destination, System.Security.Cryptography.PaddingMode paddingMode, out int bytesWritten) { throw null; }
         public bool ValidKeySize(int bitLength) { throw null; }
```
I'm kind of torn on the attributes here.  Again, for our inbox implementations it's true that CFB and ECB are unsupported... but someone could write one.

I feel more inclined to accept the attributes here than on the enum, since the enum will warn for asking the question (`if (mode == CipherMode.CFB) { throw ... }`), and in the method you're probably calling it (though saving it as a delegate for conditionally invoking would be a potentially false positive).

vcsjones replied:
I think omitting the attributes helps 0.5% of people whereas keeping the UOSP helps 99.5% of people. I find it far more likely that someone is just going to try to use the existing types than write their own implementation of a symmetric algorithm.

Also, UOSP attributes don't inherit, so if someone is writing their own implementation, they probably have their variables declared as their derived type, not the base class.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesImplementation.cs:
```diff
@@ -216,7 +221,9 @@ private static void ValidateCFBFeedbackSize(int feedback)
             {
+#pragma warning disable CA1416 // CipherMode.CFB is unsupported on Browser. It is checked elsewhere.
                 throw new CryptographicException(string.Format(SR.Cryptography_CipherModeFeedbackNotSupported, feedback, CipherMode.CFB));
+#pragma warning restore CA1416
```
This is an example of why I don't want to paint the enum members.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1031 @@
+        // Encrypts the inputBuffer into the outputBuffer using the AES encryption routine.
+        // This method writes the encrypted data into the output buffer.
+        //
+        private int EncryptData(ReadOnlySpan<byte> inputBuffer, Span<byte> outputBuffer)
```
Toss-up: Should we call them `source` and `destination` to match span style, or `inputBuffer` and `outputBuffer` to show their NetFX heritage?

eerhardt replied:
ICryptoTransform calls them `inputBuffer` and `outputBuffer`.

BasicSymmetricCipher calls them `input` and `output`.

I guess I can drop the `Buffer` part and just go with what the base class calls them.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1031 @@
+            }
+        }
+
+        private static int rot1(int val)
+        {
+            return (val << 8 & unchecked((int)0xFFFFFF00)) | (val >> 24 & unchecked((int)0x000000FF));
+        }
```
Could be replaced with `int.RotateLeft(val, 1)`

bartonjs commented on src/libraries/System.Security.Cryptography/tests/CryptoConfigTests.cs:
```diff
@@ -103,23 +103,36 @@ public static void NamedKeyedHashAlgorithmCreate(string identifier, Type actualT
-        [InlineData("System.Security.Cryptography.Rijndael", typeof(Rijndael))]
-        [InlineData("http://www.w3.org/2001/04/xmlenc#aes128-cbc", typeof(Rijndael))]
-        [InlineData("http://www.w3.org/2001/04/xmlenc#aes192-cbc", typeof(Rijndael))]
-        [InlineData("http://www.w3.org/2001/04/xmlenc#aes256-cbc", typeof(Rijndael))]
+                    yield return new object[] { "http://www.w3.org/2001/04/xmlenc#aes128-cbc", typeof(Rijndael) };
+                    yield return new object[] { "http://www.w3.org/2001/04/xmlenc#aes192-cbc", typeof(Rijndael) };
+                    yield return new object[] { "http://www.w3.org/2001/04/xmlenc#aes256-cbc", typeof(Rijndael) };
 #pragma warning restore SYSLIB0022
```
Why aren't these in the "always" case?

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/AES/AesContractTests.cs:
```diff
@@ -302,25 +303,28 @@ public static void CreateTransformExceptions()
+                    }
                 }
             }
         }
```
Since that is redundant with the tests below, disregard.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesManagedTransform.Browser.cs:
```diff
@@ -0,0 +1,1031 @@
+                }
+                if (_decryptKeyExpansion != null)
+                {
+                    Array.Clear(_decryptKeyExpansion);
```
Ah.  Um.  I guess what's there is fine.

eerhardt replied:
I'll add a comment to call this out.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SymmetricAlgorithm.cs:
```diff
@@ -262,6 +263,7 @@ public bool ValidKeySize(int bitLength)
         ///   is not a whole number of blocks.
         ///   </para>
         /// </exception>
+        [UnsupportedOSPlatform("browser")]
```
While there's not really a scenario for them, they're just math and will work fine in browser... so I don't see any reason to paint them as unsupported.

eerhardt committed:
PR feedback

Remove UnsupportedOSPlatform("browser") from CipherMode and the SymmetricAlgorithm base class. These create too many false positives and aren't technically accurate since other implementations may support these on browser.
